src/dump.c

```diff
@@ -3100,14 +3126,19 @@ static jl_value_t *_jl_restore_incremental(ios_t *f)
     // now all of the interconnects will be created
     jl_recache_types(); // make all of the types identities correct
     init_order = jl_finalize_deserializer(&s, tracee_list); // done with f and s (needs to be after recache types)
+
+    jl_array_t *extmeths = jl_alloc_vec_any(0);  // make array of references for GC
+    linkedlist_to_array(extmeths, &external_methods);
+
+    JL_GC_PUSH3(&init_order, &restored, &extmeths);
+    jl_gc_enable(en);
+
     jl_insert_methods(&external_methods); // hook up methods of external generic functions (needs to be after recache types)
     jl_recache_other(); // make all of the other objects identities correct (needs to be after insert methods)
```

Comment:
This call also re-writes the contents of the `external_methods` linked list (which is why it isn't stored in an array in the first place, since we need to ensure that the memory doesn't move during `push`). A better alternative strategy here is perhaps to build the deserialization state directly into `extmeths`. We probably just will need to pre-allocate the length (probably by calculating it during the serialization step – perhaps just collect them into an array and serializing that would work).